mapsims/noise.py

```diff
@@ -0,0 +1,194 @@
+        in microK_CMB or microK_RJ (based on return_uK_CMB)
+
+        Parameters
+        ----------
+
+        telescope : {"SA", "LA"}
+            Either LA or SA for Large and Small aperture
+        band : int
+            Detectors frequency band in GHz
+        nside : int
+            Output HEALPix NSIDE
+        return_uK_CMB : bool
+            True, output is in microK_CMB, False output is in microK_RJ
+        sensitivity_mode : str
+            Value should be threshold, baseline or goal to use predefined sensitivities
+        apply_beam_correction : bool
+            Include the effect of the beam in the noise angular power spectrum
+        apply_kludge_correction : bool
+            If True, reduce the hitcount by a factor of 0.85 to account for not-uniformity in the scanning
+        scanning_strategy : str
+            Choose between the available scanning strategy hitmaps "classical" or "opportunistic" or
+            path to another hitmap
+        LA_number_LF : int
+            Number of Low Frequency tubes in LAT
+        LA_number_MF : int
+            Number of Medium Frequency tubes in LAT
+        LA_number_UHF : int
+            Number of Ultra High Frequency tubes in LAT
+        SA_years_LF : int
+            Number of years for the Low Frequency detectors to be deployed on the Small Aperture telescopes
+        SA_one_over_f_mode : {"pessimistic", "optimistic", "none"}
+            Correlated noise performance of the detectors on the Small Aperture telescopes
+        """
+
+        self.telescope = telescope
+        assert telescope in ["LA", "SA"]
+        self.band = int(band)
+        self.sensitivity_mode = sensitivity_modes[sensitivity_mode]
+        self.apply_beam_correction = apply_beam_correction
+        self.apply_kludge_correction = apply_kludge_correction
+        self.nside = nside
+        self.return_uK_CMB = return_uK_CMB
+        self.ell_max = 3 * nside
+        self.LA_number_LF = LA_number_LF
+        self.LA_number_MF = LA_number_MF
+        self.LA_number_UHF = LA_number_UHF
+        self.SA_years_LF = SA_years_LF
+        self.SA_one_over_f_mode = one_over_f_modes[SA_one_over_f_mode]
+        self.SA_remove_kluge = SA_remove_kluge
+
+        # Load hitmap and compute sky fraction
+
+        if os.path.exists(scanning_strategy):
+            hitmap_filename = scanning_strategy
+        else:
+            hitmap_filename = data.get_pkg_data_filename(
+                "data/total_hits_{}_{}.fits.gz".format(
+                    self.telescope, scanning_strategy
+                )
+            )
+        self.hitmap = hp.ud_grade(
+            hp.read_map(hitmap_filename, verbose=False), nside_out=self.nside
+        )
+        self.hitmap /= self.hitmap.max()
+        # Discard pixels with very few hits that cause border effects
+        self.hitmap[self.hitmap < 1e-3] = 0
+        self.sky_fraction = (self.hitmap != 0).sum() / len(self.hitmap)
```

Comment:
Mmmh, what's happening here? The sky fraction you should pass to the noise calculator (if in the end you're gonna divide by `np.sqrt(self.hitmap)`) is `np.mean(self.hitmap)`, but I think you're passing the fraction of pixels with non-zero hits instead...

Reply:
@jcolinhill maybe you can help here, I compute the sky fraction to pass to `Simons_Observatory_V3_SA_noise`, it says `f_sky` is a number between 0 and 1, I assumed this is the ratio between the number of pixels observed and the total number of pixels. is this right?

Reply:
Alright @zonca @jcolinhill , here's a boring explanation of why the sky fraction there should be <Nhits>. I'm copy-pasting this from an email I exchanged with Raphael and others, since this has been a point that has confused quite a few people (in fact we're gonna change the way this is explained in the paper to make it clearer). Sorry this has been confusing.

The logic to how we generated noise maps is as follows: the white noise rms in pixel p should be inversely proportional to sqrt(Nhits_p) (where Nhits_p is our Nhits map with arbitrary normalization), so sigma_p = K/sqrt(Nhits_p * A_pix) (where A_pix is the pixel area). You can calculate the overall constant K by noting that the total hits count must be kept constant regardless of scanning strategy. So Sum_p 1/sigma_p^2 = constant.
For our Nhits map with arbitrary normalization this is:
           sum_p 1/sigma_p^2 = <Nhits>*4*pi/K^2
For a homogeneous patch with sky fraction fsky (which is what the v3 calculator returns a noise level sigma_fsky for), this is
           sum_p 1/sigma_p^2 = sum_p A_pix/sigma_fsky^2 = 4*pi * fsky / sigma_fsky^2
Equating both, you then get that you can estimate the constant K by computing sigma_fsky from the v3 calculator for fsky=<Nhits> (or you could pass <Nhits> multiplied by any arbitrary constant and then divide the answer by the sqrt of that constant).

So, what we do is, compute fsky=<Nhits>, get a noise curve from v3 for that fsky (note that in v3 the noise curves just scale with fsky), generate a homogeneous noise realization and then divide it by sqrt(Nhits).

Reply:
@zonca  sure, I'm fine with any choice as long as it's done consistently and it doesn't affect performance significantly.

A pedantic note would be that I don't think the definition you're using here is necessarily the "standard" one. For one, it depends on an arbitrary threshold of 1e-3 (so why not take pixels with 1e-6? Or use a 1e-2 threshold?). I would actually say that a more "standard" definition would be either fsky=< Nhits^2 >/< Nhits >^2 or fsky=< Nhits^2?>^2/ < Nhits >^4, which correspond to the effective number of modes for noise- and signal-dominated maps.

But as I said, as long as this is done consistently, it's all fine by me.

Reply:
thanks @damonge,
@jcolinhill which definition of `f_sky` do you assume in the `Simons_Observatory_V3_SA_noise` function? or it doesn't matter?

Reply:
thanks @jcolinhill 
I will keep `f_sky` as defined from boolean coverage, anyway this is only used internally and normalized out.
However I will implement higher level functions in `mapsims` for the 2 effective sky fraction definitions provided by @damonge